### ISSUE TITLE
Collection ABCs such as Sequence should not be assumed to be Hashable

### ISSUE THREAD

**Issue by Bruntaz** [microsoft/pyright#7430]
**Describe the bug**
Some (but not all) of the generic versions of interfaces in the `collections.abc` module are regarded as being `Hashable` when there's no guarantee that the concrete type is actually hashable. For example, `Sequence` and `Collection` are both considered `Hashable` by Pyright, but `Mapping` is not. I think the default behaviour should be that the generic collections should be considered unhashable since there's nothing stopping you from passing `[]` to a `Sequence`, which could then be considered a valid dictionary key.

**Code or Screenshots**
[Playground here](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGUkBHAVySxiQBooAJAnABbFk3ALIEKFdFgDm3AMIJkaTFgCwAKE2aAJkmBRgYMAApNUC1AIAuKPHQ5Unc5aK3UzCiJcWYthixsHM4allA6tuKS0nI%2BUEi2fILCXHHAtoqIKBjYIQCU1nEA%2BlAAvFAA3jZQAIwAvhYAxFAAKgCeFEhQAET2jgDaAKpYANZYYADuWAC63WQ4UOO0QslEyFAmcCAg2XnFZZVutQ1QzQByYPHb1BtbO2h7oZYl5RV%2Bx01QF1fgIBvSd2y3ACrHYXR0YCQCyWZEo8DaUBWQjWSEeYRelQiH1OrQ6XW6USksiGo3GU24wzGkxmcwcizAy34yPWm2o91QaOeBwqCWx50uSGuf1Z212%2B1e6T5XwFQv%2B7DZQKgmWUOSw4Uh0IZsK8CKRKUeQA)

```python
from collections.abc import Sequence, Hashable, Mapping, Collection


def foo(
    a: list,
    b: tuple,
    c: Sequence,
    d: Mapping,
    e: Hashable,
    f: Collection,
):
    _ = {a: 1}  # Type "list[Unknown]" is not hashable (correct)
    _ = {b: 1}  # No error (correct)
    _ = {c: 1}  # No error (incorrect, Sequence does not imply hashable)
    _ = {d: 1}  # Type "Mapping[Unknown, Unknown]" is not hashable (correct)
    _ = {e: 1}  # No error (correct)
    _ = {f: 1}  # No error (incorrect, Collection does not imply hashable)
```

Minimal example of this causing a type error in running code:

[Playground here](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGUkBHAVySxiQFgAoHnnGAVSokIAPoATIQQBcUCejQBtBizYcl6LKgC6AGig5UIHVAC8UAN4BfPtwlJgUAhImTpACjjUJOOatZ2JE1tfSgANwJ4OSMQAEoZHihkw0FhUXdUAiVvEF9TC0j4Oxc3KSyPDwBGAwAmOIMAIgAJJEQwRrjkgGIoACVmXChgLS5uUsyCDyUAZgMAFjCWtvgOrqhegBUATwokAFEQcBA5QYALAhwLomQoVF2kOQByeHQjJ54gA)

```python
from collections.abc import Sequence


scatter_data: dict[Sequence[int], str] = {}

def add_data(coords: Sequence[int], val: str):
    scatter_data[coords] = val

add_data((1, 2), "Hello")  # Runs fine
add_data([3, 4], "Hello")  # TypeError: unhashable type: 'list'
```

**VS Code extension or command-line**
Found using Pyright from the commandline using v1.1.350, verified in Pyright playground v1.1.353.

**Comment by erictraut** [comment]
Hashability is a weird thing in python. It doesn't follow normal inheritance rules, so wrong conclusions can be drawn by a type checker. For these reasons, I was hesitant to add support for hashability checks in pyright in the first place, but I was pushed to do so by a pyright user and contributor. 

I think the current behavior in pyright is the best we can do given the odd behavior of hashability in Python. It may produce false negatives in some cases, as you've shown.
